src/backend/gl/src/device.rs

```diff
@@ -386,11 +451,13 @@ impl d::Device<B> for Device {
     fn create_pipeline_layout<IS, IR>(&self, _: IS, _: IR) -> n::PipelineLayout
     where
         IS: IntoIterator,
-        IS::Item: Borrow<n::DescriptorSetLayout>,
+        IS::Item: Borrow<Vec<pso::DescriptorSetLayoutBinding>>,
         IR: IntoIterator,
         IR::Item: Borrow<(pso::ShaderStageFlags, Range<u32>)>,
     {
-        n::PipelineLayout
+        n::PipelineLayout {
+            desc_remap_data: Arc::new(RwLock::new(n::DescRemapData::new())),
```

Comment:
If I understand correctly, you chose to have `Arc+RwLock` here because creating a pipeline needs to modify those mappings. This is concerning, actually. Can't we have the remapping table constantly associated with the set of descriptor set layouts? In this case, it would be immutable and filled up right here.

Reply:
I'll look into this when I get back.

Reply:
Ok, I just realized something. The mapping should be owned by the pipeline and not shared between pipelines. If we have two pipelines sharing the same layout but with different shaders then we will get a suboptimal remapping.

The issue with making the pipelines own the remapping is that the pipeline doesn't get passed to `bind_graphics_descriptor_sets,` the pipeline layout does. The pipeline is never passed in the creation of the descriptors sets or their layout either. To store the remap data in it's correct place we'd either have to keep track of the current bound pipeline or pass in the pipeline to `bind_graphics_descriptor_sets`.

We can't store them with the the descriptor set layouts because the layouts are completely unrelated. The remapping is dependent on the binding points and sets set in the shaders of a single pipeline, and is only used when we bind the related descriptor sets after binding that pipeline.

Reply:
> If we have two pipelines sharing the same layout but with different shaders then we will get a suboptimal remapping.

This is fine by me. Pipeline layout *is* that mapping, semantically speaking. We shouldn't go out of our way to make it more efficient than the API suggests.

Reply:
I think this is still not the way I imagine it should be. The pipeline layout should not have internal mutability. When it's created, it has all the information required to map from descriptor binding/sets to GL resource slots, and that's the mapping it should carry. All the actual pipeline construction should *use* that mapping but not modify it.